exercise/api/views.py

```diff
@@ -222,6 +226,77 @@ def create(self, request, exercise_id, version):
+        methods=['post'],
+    )
+    def submit(self, request, *args, **kwargs):
+        """Submit a new solution to the exercise for grading.
+        Students are allowed to submit via this endpoint. In a group submission,
+        the student group ID is defined with the POST parameter _aplus_group.
+        The POST data is used as the submission data and files may be uploaded too.
+        """
+        # Stop submit trials for e.g. chapters.
+        # However, allow posts from exercises switched to maintenance status.
+        if not self.exercise.is_submittable:
+            return self.http_method_not_allowed(request, *args, **kwargs)
+
+        data = None
+        status_code = None
+        headers = None
+        submission_status, issues, students = (
+            self.exercise.check_submission_allowed(request.user.userprofile, request)
+        )
+        if submission_status == self.exercise.SUBMIT_STATUS.ALLOWED:
+            new_submission = Submission.objects.create_from_post(
+                self.exercise, students, request)
+            if new_submission:
+                page = self.exercise.grade(request, new_submission, use_messages=False)
+
+                # Enroll after succesfully submitting to the enrollment exercise.
+                if self.exercise.status in (
+                    LearningObject.STATUS.ENROLLMENT,
+                    LearningObject.STATUS.ENROLLMENT_EXTERNAL,
+                ) and new_submission.status == Submission.STATUS.READY:
+                    self.instance.enroll_student(request.user)
+
+                status_code = status.HTTP_201_CREATED
+                headers = {
+                    'Location': reverse('api:submission-detail', kwargs={
+                                        'submission_id': new_submission.id,
+                                        }, request=request),
+                }
+                if page.errors:
+                    data = {'errors': page.errors}
+            else:
+                data = {
+                    'detail': _("The submission could not be saved for some reason. "
+                                "The submission was not registered."),
+                }
+                status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
+        else:
+            data = {'errors': issues}
+            status_code = status.HTTP_400_BAD_REQUEST
+
+        return Response(data, status=status_code, headers=headers)
+
+    def get_access_mode(self):
+        # The API is not supposed to use the access mode permission in views,
+        # but this is currently required so that enrollment exercises work in
+        # the CourseVisiblePermission.
+        access_mode = ACCESS.STUDENT
+
+        # Loosen the access mode if this is an enrollment exercise.
+        if self.exercise.status in (
+                LearningObject.STATUS.ENROLLMENT,
+                LearningObject.STATUS.ENROLLMENT_EXTERNAL,
+              ):
+            access_mode = ACCESS.ENROLL
```

Comment:
This access mode makes life so hard. I hope that one day enrollment questionaire can be done differently, so that we can create enrollment object, with a flag that tells if the questionare is completed, and then exercise permissions can check if the student has enrolled or not. I think this part of the access mode is thea reason, I couldn't get rid of access modes, when I introduced the permission classes.

Reply:
I agree that is is annoying. For the moment, we have to use the `get_access_mode` method.